src/units/unit.cpp

```diff
@@ -1808,22 +1804,37 @@ int unit::resistance_ability(unit_ability_list resistance_abilities, const std::
 	return res;
 }
 
-int unit::resistance_against(const std::string& damage_name,bool attacker,const map_location& loc, const_attack_ptr weapon, const_attack_ptr opp_weapon) const
+static bool resistance_filter_matches_base(const config& cfg, bool attacker)
 {
-	std::pair<std::string, std::string> types;
-	if(opp_weapon){
-		types = opp_weapon->damage_type();
-	} else{
-		types.first = damage_name;
+	if(!(cfg["active_on"].empty() || (attacker && cfg["active_on"] == "offense") || (!attacker && cfg["active_on"] == "defense"))) {
+		return false;
 	}
 
+	return true;
+}
+
+int unit::resistance_against(const std::string& damage_name,bool attacker,const map_location& loc, const_attack_ptr weapon, const_attack_ptr opp_weapon) const
+{
 	unit_ability_list resistance_abilities = get_abilities_weapons("resistance",loc, weapon, opp_weapon);
-	int res = resistance_ability(resistance_abilities, types.first, attacker);
-	if(!(types.second).empty()){
-		res = std::max(res , resistance_ability(resistance_abilities, types.second, attacker));
+	utils::erase_if(resistance_abilities, [&](const unit_ability& i) {
+		return !resistance_filter_matches_base(*i.ability_cfg, attacker);
+	});
+	if(opp_weapon){
+		unit_ability_list abil_list = opp_weapon->get_specials_and_abilities("damage_type");
+		if(abil_list.empty()){
+			return resistance_ability(resistance_abilities, damage_name);
+		}
+		std::optional<std::string> replacement_type = opp_weapon->select_damage_type(abil_list, "replacement_type", false, resistance_abilities);
+		std::string type_damage = replacement_type.value_or(damage_name);
+		int res = resistance_ability(resistance_abilities, type_damage);
+		for(auto& i : abil_list) {
+			if(!(*i.ability_cfg)["alternative_type"].str().empty()){
```

Comment:
i will not replace type() by an alternative_type empty just because attribute is present.

Reply:
You’re welcome to implement validation of the damage type but just ignoring the empty string is not useful.

Reply:
if string is empty then is can't be a damage type valid and is more useful for me to ignore empty string what to call resistance_ability then what alternative_type is present but without string value.

Reply:
alternative_type key can't be valid without name of type valid or not elese resistance_ability will checked a ghost type and return 100

Reply:
Maybe an empty string is not a valid damage type but the point is that there are plenty more. Either you do proper validation or not. Some half measure is just going to confuse the user and code reader.

If you want to do validation naturally you need to take user defined damage types into account.

Reply:
How ta

> Maybe an empty string is not a valid damage type but the point is that there are plenty more. Either you do proper validation or not. Some half measure is just going to confuse the user and code reader.
> 
> If you want to do validation naturally you need to take user defined damage types into account.

How doing that in c++ when damage type are used in cfg files. Other type exist outside mainline. You suggest something too complex. This half mesure is the one possible then stop with that!

Reply:
I just tested and a damage type of empty works perfectly fine. So either we implement a check that only allows damage types of defined unit types or we allow anything. I'm against an arbitrary empty check unless we want to forbid an empty damage type in general. This is not the place to decide that though.

Reply:
> I just tested and a damage type of empty works perfectly fine. So either we implement a check that only allows damage types of defined unit types or we allow anything. I'm against an arbitrary empty check unless we want to forbid an empty damage type in general. This is not the place to decide that though.

So let's take this case following a unit attacking another, the first has the original type blade and a secondary type say arcane provided by damage_type and another [damage_type]alternative_type which is not specified, the second unit has a resistance of 50% at blase, 10% at arcane and the empty attribute would be if I apply your logic of 0% instead of being ignored, knowing that it is the type which does the most damage which is selected, this would mean that it is the ghost type which is taken into account and the non-existent becomes a computer ghost (it acts even though it is not supposed to exist). is using empty instead of has_attribute still useless?

Reply:
I have no real idea what you're saying. If someone adds weird damage types to their units then it's probably going to behave weird. An empty check will not change that. If they add an attack with type " " it's also going to look weird.

Again either the engine does actual proper checks or none. Anything in between is just confusing.

Reply:
there is still a difference between adding a new type of damage and forgetting to enter an attribute, if the type does not exist in mainline it can exist elsewhere and could be valid whereas an empty string cannot be valid , and checking if the type is valid would be too cumbersome since precisely any type 'electric explosive' or anything else can be legitimate in an extension and remains usable even against a mainline unit (so it's 0%) and I don't see how to differentiate a correctly written type from a writing error when we can differentiate between an attribute entered or not. You must understand that there are cases where checking that the attribute is entered is just as important as checking its simple presence and that is the case here. Here you have to make a compromise so as not to call a function unintentionally.

Reply:
I just told you that the empty string is in actual fact valid right now. You can try it yourself if you like. Is it odd to use it? Maybe. So is using a single space or a non-breaking space or a control character or a million other things.

> You must understand that there are cases where checking that the attribute is entered is just as important as checking its simple presence and that is the case here.

Hence why the code should check for its presence and not if it has a specific value.

> Here you have to make a compromise so as not to call a function unintentionally.

You are guessing at the intention of the author. Simply acting like an attribute that is clearly there but does not have a value you like is in fact not there is confusing. If you truly want to help the WML author to avoid a mistake then emit a warning. Just silently pretending like you know better what the author meant is not really helpful.

Reply:
has_attribute added